cluster-api/src/main/java/io/aiven/klaw/clusterapi/services/ConsumerGroupService.java

```diff
@@ -0,0 +1,257 @@
+package io.aiven.klaw.clusterapi.services;
+
+import io.aiven.klaw.clusterapi.models.ApiResponse;
+import io.aiven.klaw.clusterapi.models.consumergroup.OffsetDetails;
+import io.aiven.klaw.clusterapi.models.consumergroup.OffsetResetType;
+import io.aiven.klaw.clusterapi.models.consumergroup.OffsetsTiming;
+import io.aiven.klaw.clusterapi.models.consumergroup.ResetConsumerGroupOffsetsRequest;
+import io.aiven.klaw.clusterapi.models.enums.KafkaSupportedProtocol;
+import io.aiven.klaw.clusterapi.utils.ClusterApiUtils;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.ExecutionException;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.AlterConsumerGroupOffsetsOptions;
+import org.apache.kafka.clients.admin.AlterConsumerGroupOffsetsResult;
+import org.apache.kafka.clients.admin.DescribeTopicsResult;
+import org.apache.kafka.clients.admin.ListOffsetsResult;
+import org.apache.kafka.clients.admin.OffsetSpec;
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.TopicPartitionInfo;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class ConsumerGroupService {
+  private final ClusterApiUtils clusterApiUtils;
+
+  public ConsumerGroupService(ClusterApiUtils clusterApiUtils) {
+    this.clusterApiUtils = clusterApiUtils;
+  }
+
+  public ApiResponse resetConsumerGroupOffsets(
+      String environment,
+      KafkaSupportedProtocol protocol,
+      String clusterIdentification,
+      ResetConsumerGroupOffsetsRequest consumerGroupOffsetsRequest)
+      throws Exception {
+    log.info(
+        "Reset consumer group offsets env {} request {}", environment, consumerGroupOffsetsRequest);
+
+    if (OffsetResetType.TO_DATE_TIME.equals(consumerGroupOffsetsRequest.getOffsetResetType())
+        && null == consumerGroupOffsetsRequest.getConsumerGroupResetTimestampMilliSecs()) {
+      throw new Exception("Timestamp must be provided for reset type TO_DATE_TIME");
```

Comment:
can we make this exception return a 400 bad request? similar to how we have done it in core?

Reply:
In cluster api (in all controllers), we are throwing only exceptions as it is always caught by core api, and finally FE gets a standard error message. 
We can change this all over but also need some changes in core so that this bad request is propagated to FE as well.

Reply:
I think that would be something we should look at long term.